crates/ruff_python_semantic/src/module.rs

```diff
@@ -0,0 +1,284 @@
+use std::fmt::Formatter;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use ruff_db::file_system::FileSystemPath;
+use ruff_db::vfs::{VfsFile, VfsPath};
+
+use crate::Db;
+
+pub mod resolver;
+
+/// A module name, e.g. `foo.bar`.
+///
+/// Always normalized to the absolute form (never a relative module name, i.e., never `.foo`).
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct ModuleName(smol_str::SmolStr);
+
+impl ModuleName {
+    #[inline]
+    pub fn new(name: &str) -> Self {
+        assert!(!name.is_empty());
+        assert!(!name.starts_with('.'), "module name must be absolute");
+        assert!(!name.ends_with('.'), "module cannot end with a '.'");
+
+        Self(smol_str::SmolStr::new(name))
+    }
+
+    /// An iterator over the components of the module name:
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ruff_python_semantic::module::ModuleName;
+    ///
+    /// assert_eq!(ModuleName::new("foo.bar.baz").components().collect::<Vec<_>>(), vec!["foo", "bar", "baz"]);
+    /// ```
```

Comment:
I converted the examples to doctests and found bugs in both `parent` and `starts_with` :hand_over_mouth: